### PR TITLE
Prepare MSI installer using Pyinstaller and WIX Toolset

### PR DESCRIPTION
**Pull request by YingChen1996**
# Description
msi installer
image

directory structure after installing
image

app in "Programs and Features" / "services"
image

image

pf command in CLI
image

system path environment
image

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: build_scripts/windows/product.wxs [hunk 1]
```diff
@@ -97,6 +109,32 @@
...
+                       Guid="$(var.PromptflowCliVersionGuid)">
+                <RegistryValue Root="HKLM"
+                               Key="Software\Microsoft\$(var.ProductName)"
+                               Name="version"
+                               Type="string"
+                               Value="$(var.ProductVersion)"
+                               KeyPath="yes"/>
+            </Component>
+            <Component Id="PromptflowService" Directory="DynamicCliDir" Guid="$(var.PromptflowServiceGuid)">
+                <File Id="PromptflowServiceEXE" Source="scripts\dist\promptflow\promptflow_service.exe" KeyPath="yes" />
```
Comment (wangchao1230) [inline build_scripts/windows/product.wxs:133]: let's just name this pfs

File: build_scripts/windows/product.wxs [hunk 2]
```diff
@@ -97,6 +109,32 @@
...
+                <ServiceInstall Id="ServiceInstaller" Name="PromptflowService" DisplayName="Promptflow Service" Description="This is a promptflow service" Start="auto" Type="ownProcess" ErrorControl="normal" />
```
Comment (wangchao1230) [inline build_scripts/windows/product.wxs:134]: promptflow -> prompt flow

Comment (wangchao1230) [reply build_scripts/windows/product.wxs:134]: DisplayName: Prompt Flow Service
Description:  Provides run and connection management for prompt flow

File: build_scripts/windows/resources/NOTICE.txt [hunk 3]
```diff
@@ -0,0 +1,530 @@
+NOTICES AND INFORMATION
```
Comment (wangchao1230) [inline build_scripts/windows/resources/NOTICE.txt:1]: we already have this file in src/promptlow/NOTICE.txt, could we reuse that file?

File: build_scripts/windows/scripts/promptflow_service.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    def run(self):
+        """Main service loop. This is where work is done!"""
+        self.running = True
+        while self.running:
+            main()  # Important work
+            servicemanager.LogInfoMsg("Service running...")
+
+
+class MyServiceFramework(win32serviceutil.ServiceFramework):
```
Comment (wangchao1230) [inline build_scripts/windows/scripts/promptflow_service.py:26]: @lalala123123  FYI

File: build_scripts/windows/product.wxs [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        <?else ?>
+        <Condition Message="A newer version of $(var.ProductName) is already installed.">NOT WIX_DOWNGRADE_DETECTED</Condition>
+        <?endif ?>
+
+        <Media Id="1" Cabinet="promptflow.cab" EmbedCab="yes" CompressionLevel="high" />
+
+        <Icon Id="PromptflowIcon" SourceFile="$(var.ProductResources)logo32.ico" />
+
+        <Property Id="ARPPRODUCTICON" Value="PromptflowIcon" />
+        <Property Id="ARPHELPLINK" Value="https://learn.microsoft.com/en-us/azure/machine-learning/prompt-flow/overview-what-is-prompt-flow?view=azureml-api-2" />
```
Comment (wangchao1230) [inline build_scripts/windows/product.wxs:61]: Let's use the OSS doc link here

File: build_scripts/windows/product.wxs [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+        <CustomActionRef Id="WixBroadcastEnvironmentChange" />
+
+            <!-- User Interface -->
+        <WixVariable Id="WixUILicenseRtf" Value="$(var.ProductResources)CLI_LICENSE.rtf"/>
+
+        <UIRef Id="WixUI_Minimal"/>
+        <UIRef Id="WixUI_ErrorProgressText"/>
+
+        <!-- Show message to restart any terminals only if the PATH is changed -->
+        <CustomAction Id="Set_WIXUI_EXITDIALOGOPTIONALTEXT" Property="WIXUI_EXITDIALOGOPTIONALTEXT" Value="Please close and reopen any active terminal window to use Promptflow." />
```
Comment (wangchao1230) [inline build_scripts/windows/product.wxs:85]: use prompt flow to replace Promptflow

File: build_scripts/windows/product.wxs [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        <Media Id="1" Cabinet="promptflow.cab" EmbedCab="yes" CompressionLevel="high" />
+
+        <Icon Id="PromptflowIcon" SourceFile="$(var.ProductResources)logo32.ico" />
+
+        <Property Id="ARPPRODUCTICON" Value="PromptflowIcon" />
+        <Property Id="ARPHELPLINK" Value="https://learn.microsoft.com/en-us/azure/machine-learning/prompt-flow/overview-what-is-prompt-flow?view=azureml-api-2" />
+        <Property Id="ARPURLINFOABOUT" Value="https://learn.microsoft.com/en-us/azure/machine-learning/prompt-flow/overview-what-is-prompt-flow?view=azureml-api-2" />
+        <Property Id="ARPURLUPDATEINFO" Value="https://learn.microsoft.com/en-us/azure/machine-learning/prompt-flow/overview-what-is-prompt-flow?view=azureml-api-2" />
+        <Property Id="MSIFASTINSTALL" Value="7" />
+        <Property Id="ApplicationFolderName" Value="Promptflow" />
```
Comment (wangchao1230) [inline build_scripts/windows/product.wxs:65]: what's this folder? why need to cap the first character?

File: build_scripts/windows/promptflow.wixproj [hunk 8]
```diff
@@ -0,0 +1,73 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="4.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <!-- Project -->
+  <PropertyGroup>
+    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
+    <Platform Condition=" '$(Platform)' == '' ">x86</Platform>
+    <ProductVersion>3.10</ProductVersion>
+    <ProjectGuid>04ff6707-750d-4474-89b3-7922c84721be</ProjectGuid>
+    <SchemaVersion>2.0</SchemaVersion>
+    <OutputName>Promptflow</OutputName>
```
Comment (wangchao1230) [inline build_scripts/windows/promptflow.wixproj:10]: Let's use all lower case for file/directory names: promptflow

File: build_scripts/windows/product.wxs (hunk 7 above)
Comment (YingChen1996) [reply build_scripts/windows/product.wxs:65]: folder when install msi: 
image

File: build_scripts/windows/product.wxs (hunk 5 above)
Comment (YingChen1996) [reply build_scripts/windows/product.wxs]: updated

File: build_scripts/windows/product.wxs (hunk 6 above)
Comment (YingChen1996) [reply build_scripts/windows/product.wxs]: updated

File: build_scripts/windows/product.wxs (hunk 1 above)
Comment (YingChen1996) [reply build_scripts/windows/product.wxs]: updated

File: build_scripts/windows/promptflow.wixproj (hunk 8 above)
Comment (YingChen1996) [reply build_scripts/windows/promptflow.wixproj]: done

File: build_scripts/windows/product.wxs (hunk 7 above)
Comment (YingChen1996) [reply build_scripts/windows/product.wxs]: have changed to lower case: promptflow

File: build_scripts/windows/README.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
+Building the Windows MSI Installer
```
Comment (wangchao1230) [inline build_scripts/windows/README.md:1]: shall we put this build_scripts under scripts folder? 
kind of confuse as those scripts are also build scripts

File: build_scripts/windows/README.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+This document provides instructions on creating the MSI installer.
+
+Prerequisites
+-------------
+
+1. Turn on the '.NET Framework 3.5' Windows Feature (required for WIX Toolset).
+2. Install 'Microsoft Build Tools 2015'.
+    https://www.microsoft.com/download/details.aspx?id=48159
+3. You need to have curl.exe, unzip.exe and msbuild.exe available under PATH.
+4. Install 'WIX Toolset build tools' following the instructions below.
```
Comment (wangchao1230) [inline build_scripts/windows/README.md:13]: make these steps as a ps1 file, so we will be easier to create a workflow from it

File: build_scripts/windows/README.md [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+   - `unzip wix-archive.zip` and `del wix-archive.zip`
+5. We recommend creating a clean virtual Python environment and installing all dependencies using src/promptflow/setup.py.
+   - `python -m venv venv`
+   - `venv\Scripts\activate`
+   - `pip install promptflow[azure,executable] promptflow-tools`
+
+
+Building
+--------
+1. `cd build_scripts/windows/scripts` and run `pyinstaller promptflow.spec`.
```
Comment (wangchao1230) [inline build_scripts/windows/README.md:25]: add icon also for the built pf.exe

File: build_scripts/windows/scripts/promptflow.spec [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+datas += collect_data_files('streamlit')
+datas += copy_metadata('streamlit')
+datas += collect_data_files('streamlit_quill')
+hidden_imports = ['streamlit.runtime.scriptrunner.magic_funcs']
+
+service_hidden_imports = ['win32timezone']
+
+block_cipher = None
+
+pf_a = Analysis(
```
Comment (wangchao1230) [inline build_scripts/windows/scripts/promptflow.spec:15]: Could we add more metas for exe?
image

File: build_scripts/windows/scripts/promptflow.spec [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+from PyInstaller.utils.hooks import collect_data_files
+from PyInstaller.utils.hooks import copy_metadata
+
+datas = [('../resources/CLI_LICENSE.rtf', '.'), ('../../../src/promptflow/NOTICE.txt', '.')]
+datas += collect_data_files('streamlit')
+datas += copy_metadata('streamlit')
+datas += collect_data_files('streamlit_quill')
+hidden_imports = ['streamlit.runtime.scriptrunner.magic_funcs']
+
+service_hidden_imports = ['win32timezone']
```
Comment (wangchao1230) [inline build_scripts/windows/scripts/promptflow.spec:11]: which part requires this? pfs?

File: build_scripts/windows/README.md (hunk 9 above)
Comment (YingChen1996) [reply build_scripts/windows/README.md]: move windows folder to scripts

File: build_scripts/windows/scripts/promptflow.spec (hunk 13 above)
Comment (YingChen1996) [reply build_scripts/windows/scripts/promptflow.spec]: yeah

windows service will use win32timezone

File: build_scripts/windows/README.md (hunk 11 above)
Comment (YingChen1996) [reply build_scripts/windows/README.md]: done

File: scripts/windows/README.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+   - `python -m venv venv`
+   - `venv\Scripts\activate`
+   - `pip install promptflow[azure,executable] promptflow-tools`
+
+
+Building
+--------
+1. `cd scripts/windows/scripts` and run `pyinstaller promptflow.spec`.
+2. `cd scripts/windows` and Run `msbuild /t:rebuild /p:Configuration=Release /p:Platform=x64 promptflow.wixproj`.
+3. The unsigned MSI will be in the `scripts/windows/out` folder.
```
Comment (wangchao1230) [inline scripts/windows/README.md:27]: please add gitignore to the folders not expected to checkin like: out, wix

File: scripts/windows/README.md [hunk 15]
```diff
@@ -11,7 +11,7 @@ Prerequisites
     https://www.microsoft.com/download/details.aspx?id=48159
 3. You need to have curl.exe, unzip.exe and msbuild.exe available under PATH.
 4. Install 'WIX Toolset build tools' following the instructions below.
-   - Enter the directory where the README is located (`cd build_scripts/windows`), `mkdir wix` and `cd wix`.
+   - Enter the directory where the README is located (`cd scripts/windows`), `mkdir wix` and `cd wix`.
```
Comment (wangchao1230) [inline scripts/windows/README.md:14]: maybe do this folder structure: scripts/installer/windows

File: scripts/windows/scripts/promptflow.spec [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+    pathex=[],
+    binaries=[],
+    datas=datas,
+    hiddenimports=hidden_imports,
+    hookspath=[],
+    hooksconfig={},
+    runtime_hooks=[],
+    excludes=[],
+    win_no_prefer_redirects=False,
+    win_private_assemblies=False,
```
Comment (wangchao1230) [inline scripts/windows/scripts/promptflow.spec:26]: could add a version parameter here.

Comment (wangchao1230) [reply scripts/windows/scripts/promptflow.spec:26]: learn more: https://github.com/pyinstaller/pyinstaller/blob/develop/PyInstaller/building/api.py

File: build_scripts/windows/scripts/promptflow.spec (hunk 12 above)
Comment (YingChen1996) [reply build_scripts/windows/scripts/promptflow.spec]: image

File: scripts/installer/windows/scripts/version_info.txt [hunk 17]
```diff
@@ -0,0 +1,42 @@
+# UTF-8
+#
+# For more details about fixed file info 'ffi' see:
+# http://msdn.microsoft.com/en-us/library/ms646997.aspx
+VSVersionInfo(
+  ffi=FixedFileInfo(
+    # filevers and prodvers should be always a tuple with four items: (1, 2, 3, 4)
+    # Set not needed items to zero 0.
+    filevers=(0, 0, 0, 1),
```
Comment (wangchao1230) [inline scripts/installer/windows/scripts/version_info.txt:9]: use 1.0.0.0 as default one?

File: build_scripts/windows/README.md (hunk 10 above)
Comment (YingChen1996) [reply build_scripts/windows/README.md]: leave it to the next pr

**Review by wangchao1230 — DISMISSED** [review]

**Review by lalala123123 — APPROVED** [review]

**Review by wangchao1230 — APPROVED** [review]

**Review by lalala123123 — APPROVED** [review]

**Pull request merged**